Gems/Atom/RHI/Vulkan/Code/Source/RHI/Instance.cpp

```diff
@@ -286,5 +288,66 @@ namespace AZ
             }
             return supportedDevices;
         }
+
+        void Instance::CreateDebugMessenger()
+        {
+            auto validationMode = GetValidationMode();
+            if (validationMode != RHI::ValidationMode::Disabled)
+            {
+                auto messagesTypeMask =
+                    Debug::DebugMessageTypeFlag::Error | Debug::DebugMessageTypeFlag::Warning | Debug::DebugMessageTypeFlag::Performance;
+                if (validationMode == RHI::ValidationMode::Verbose)
+                {
+                    messagesTypeMask |= Debug::DebugMessageTypeFlag::Debug | Debug::DebugMessageTypeFlag::Info;
+                }
+
+                Debug::InitDebugMessages(m_instance, messagesTypeMask);
+            }
+        }
+
+        void Instance::UpdateNativeInstance(RHI::XRRenderingInterface* xrSystem)
+        {
+            if (m_isXRInstanceCreated)
+            {
+                AZ_Warning("Vulkan", false, "XR Vulkan instance is already created");
+                return;
+            }
+
+            RHI::Ptr<XRInstanceDescriptor> xrInstanceDescriptor = aznew XRInstanceDescriptor();
+            xrInstanceDescriptor->m_inputData.m_createInfo = &m_instanceCreateInfo;
+
+            // Init the new native instance for XR
+            AZ::RHI::ResultCode result = xrSystem->InitNativeInstance(xrInstanceDescriptor.get());
+            AZ_Warning("Vulkan", result == RHI::ResultCode::Success, "Xr instance creation was not successfull");
+            if (result == RHI::ResultCode::Success)
+            {
+                //Delete existing VkInstance
+                ShutdownInstance();
+
+                //Update the native object from the passed by the XR module
+                m_instance = xrInstanceDescriptor->m_outputData.m_xrVkInstance;
+
+                //Re-add support for validation with the updated VkInstance
+                CreateDebugMessenger();
+
+                //Get number of  Physical devices
+                uint32_t numPhysicalDevices = xrSystem->GetNumPhysicalDevices();
+                    
+                //Clear any existing physical devices
+                m_supportedDevices.clear();
+
+                //Re-populate physical devices from XR module
+                for (uint32_t i = 0; i < numPhysicalDevices; i++)
```

Comment:
reserve in advance to avoid memory reallocations of the container inside the loop.

````
m_supportedDevices.clear();
m_supportedDevices.reserve(numPhysicalDevices);
````